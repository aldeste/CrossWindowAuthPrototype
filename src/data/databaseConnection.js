// @flow
import Sequelize from "sequelize";
import {
  DB_USER,
  DB_PASS,
  DB_NAME,
  CONNECTION_SETTINGS
} from "../../config/config";

const connection = new Sequelize(DB_NAME, DB_USER, DB_PASS, {
  logging: false,
  define: {
    underscored: false,
    freezeTableName: false,
    syncOnAssociation: true,
    charset: "utf8mb4",
    collate: "utf8mb4_general_ci",
    timestamps: true
  },
  ...CONNECTION_SETTINGS
});

export default connection;
