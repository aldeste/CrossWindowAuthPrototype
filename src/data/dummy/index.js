// @flow
import filmsDummy from "../../data/dummy/films.swapi";
import peopleDummy from "../../data/dummy/people.swapi";
import planetsDummy from "../../data/dummy/planets.swapi";
import speciesDummy from "../../data/dummy/species.swapi";
import starshipsDummy from "../../data/dummy/starships.swapi";
import vehiclesDummy from "../../data/dummy/vehicles.swapi";

export {
  filmsDummy,
  peopleDummy,
  planetsDummy,
  speciesDummy,
  starshipsDummy,
  vehiclesDummy
};
