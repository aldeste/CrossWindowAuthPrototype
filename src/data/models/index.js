export Film from "./film";
export Person from "./person";
export Planet from "./planet";
export Species from "./species";
export Starship from "./starship";
export Vehicle from "./vehicle";
