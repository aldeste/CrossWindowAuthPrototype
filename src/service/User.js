import { type DataLoaders } from "../schema/apiHelper";

export type Viewer = {
  birthYear: string,
  created: date,
  edited: date,
  eyeColor: string,
  gender: string,
  hairColor: string,
  height: number,
  id: string,
  mass: number,
  name: string,
  personId?: ?number,
  skinColor: string,
  token: string,
  GraphQLType: string
};

export default class User {
  birthYear: string;
  created: date;
  edited: date;
  eyeColor: string;
  gender: string;
  hairColor: string;
  height: number;
  id: string;
  mass: number;
  name: string;
  skinColor: string;
  token: string;
  GraphQLType: string;
  homeworld: number;

  constructor(data: Viewer) {
    this.birthYear = data.birthYear;
    this.created = data.createdAt;
    this.edited = data.updatedAt;
    this.eyeColor = data.eyeColor;
    this.gender = data.gender;
    this.hairColor = data.hairColor;
    this.height = data.height;
    this.id = data.id;
    this.mass = data.mass;
    this.name = data.name;
    this.skinColor = data.skinColor;
    this.token = data.token;

    this.homeworld = data.homeworld.id;

    this.GraphQLType = data.GraphQLType;
  }

  static async gen(
    viewer: ?Viewer,
    id: string,
    { Person }: DataLoaders
  ): Promise<?Object> {
    const data: Object | null = await Person.load(id);
    return data ? new User(data) : null;
  }

  static async genMany(
    viewer: ?Viewer,
    { Person }: DataLoaders,
    connection: Class<*>,
    ids: Array<string>
  ): Promise<?Object> {
    const data: Object | null = await connection.findAll({
      where: { id: [ids] }
    });

    await data.forEach(result => Person.prime(result.id, result.dataValues));

    return await data.filter(result => result).map(result => new User(result));
  }
}
