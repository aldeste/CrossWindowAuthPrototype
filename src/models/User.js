// @flow
import { type DataLoaders } from "../schema/apiHelper";
import InfoFields, {
  type InfoFieldFromDatabase,
  type InfoFieldTypes
} from "./InfoFields";
import { Person as dbPerson } from "../data";

export type Viewer = InfoFieldTypes & {
  birthYear: string,
  eyeColor: string,
  gender: string,
  hairColor: string,
  height: number,
  mass: number,
  name: string,
  homeworld: string,
  personId?: ?number,
  skinColor: string,
  token: string
};

type PersonModel = InfoFieldFromDatabase & {
  birthYear: string,
  eyeColor: string,
  gender: string,
  hairColor: string,
  height: number,
  mass: number,
  name: string,
  homeworld: string,
  skinColor: string,
  token: string,
  homeworld: {
    id: number
  }
};

export default class UserInstance extends InfoFields {
  birthYear: string;
  eyeColor: string;
  gender: string;
  hairColor: string;
  height: number;
  mass: number;
  name: string;
  skinColor: string;
  token: string;
  homeworld: number;

  constructor(data: PersonModel) {
    super(data);
    this.birthYear = data.birthYear;
    this.eyeColor = data.eyeColor;
    this.gender = data.gender;
    this.hairColor = data.hairColor;
    this.height = data.height;
    this.mass = data.mass;
    this.name = data.name;
    this.skinColor = data.skinColor;
    this.token = data.token;
    this.homeworld = data.homeworld.id;
  }

  static async gen(
    viewer: ?Viewer,
    id: string,
    { Person }: DataLoaders
  ): Promise<?UserInstance> {
    const data: PersonModel | null = await Person.load(id);
    return data ? new UserInstance(data) : null;
  }

  static async genMany(
    viewer: ?Viewer,
    ids: Array<string>,
    { Person }: DataLoaders
  ): Promise<?Array<UserInstance>> {
    const data: Array<PersonModel> | null = await dbPerson
      .scope("withIds")
      .findAll({ where: { id: [ids] } })
      // Only get pure JSON from fetch,
      // Though unlikely, this prevents memory leak.
      .map(response => response.toJSON());

    // return imideately if failed to fetch
    if (!data) return null;

    // We put each field in the dataloader cache
    await data.forEach(result => Person.prime(result.id, result));

    return await data.map(result => new UserInstance(result));
  }
}
