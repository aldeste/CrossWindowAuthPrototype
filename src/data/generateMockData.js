// @flow
import chalk from "chalk";
import { toGlobalId } from "graphql-relay";
import { Person, Planet } from "../data";
import { peopleDummy, planetsDummy } from "./dummy";
import ProgressBar from "progress";

// Generates data for the database.
export default async function generateMockData(
  forceInsert: boolean = false,
  useProgressBar: boolean = false
): Promise<void> {
  // Force insert based on boolean, or if the database is empty
  if (forceInsert || (await Person.count()) === 0) {
    // Some fields in the mock data are listed as unknown, which will
    // throw when inserting to database. To handle this, each method which
    // could contain an 'unknown' value is passed through this function.
    const isKnown = (val: mixed | null) => (val === "unknown" ? null : val);

    // Padding in the console helps readability
    console.log();

    // Initiallize progress bar if progress bar option is set
    const bar: null | ProgressBar =
      useProgressBar &&
      new ProgressBar("Initiating database :bar :percent :current/:total", {
        complete: "█",
        incomplete: "░",
        clear: true,
        total:
          peopleDummy.length * 2 +
            planetsDummy.length +
            planetsDummy.reduce((p, c) => [...p, ...c.residents], []).length
      });

    // Fill the database with dymmy people
    await Promise.all(
      peopleDummy.map((person: Object, index: number) => {
        // Update progress bar with one tick
        bar && bar.tick(1);
        return Person.create({
          ...person,
          // Set ID to null, since it's inconsistent in the mock data.
          // We'll still be using it futher down to determine connections.
          id: null,
          token: toGlobalId("people", (index + 1).toString()),
          password: "password",
          hairColor: person.hair_color,
          mass: isKnown(person.mass) && parseInt(person.mass, 10),
          height: isKnown(person.height),
          skinColor: person.skin_color,
          eyeColor: person.eye_color,
          birthYear: person.birth_year
        });
      })
    );

    // Fill the database with dymmy planets, similar to what we did with people
    await Promise.all(
      planetsDummy.map((planet: Object) => {
        bar && bar.tick(1);
        return Planet.create({
          ...planet,
          id: null,
          climates: planet.climate,
          population:
            isKnown(planet.population) && parseInt(planet.population, 10),
          terrains: planet.terrain,
          diameter: isKnown(planet.diameter),
          orbitalPeriod: isKnown(planet.orbital_period),
          rotationPeriod: isKnown(planet.rotation_period),
          surfaceWater: isKnown(planet.surface_water)
        });
      })
    );

    // Associate homeworld to each user
    await Promise.all(
      // Loop through each dummy user
      peopleDummy.map(async ({ name, homeworld: home }, index) => {
        // Update progressbar
        bar && bar.tick(1);

        // Find current dummy user in databse
        const currentPerson = await Person.findOne({ where: { name } });

        // Find planet associated with dummy user, based on planet name
        const planet = planetsDummy.find(planet => planet.id === home);

        // Find planet in database, based on dummy planet's name
        const homeworld =
          planet && (await Planet.findOne({ where: { name: planet.name } }));

        // Assign homeworld to user in database
        return currentPerson.setHomeworld(homeworld);
      })
    );

    // Associate habitants to each planet
    await Promise.all(
      planetsDummy.map(async ({ name, residents }, index) => {
        // Update progressbar
        bar && bar.tick(residents.length);

        // Return imediately if planet has no residents
        if (!residents.length) return null;

        // Find current mock planet in database
        const currentPlanet = await Planet.findOne({ where: { name } });

        // Find each person associated to planet mock data
        const planetsResidents = peopleDummy
          .filter(person => residents.includes(person.id))
          .map(resident => resident.name);

        // Find all residents in database
        const residentsInDb = await Person.findAll({
          where: { name: planetsResidents }
        });

        // Add residents to planet in database
        return currentPlanet.addResidents(residentsInDb);
      })
    );

    // Notify console that each element is inserted
    return console.log(
      chalk.green.bold(
        "All fields and connections have been inserted in database"
      )
    );
  }
  // Notify console that each element already is in place
  return console.log(chalk.yellow.bold("Fields already in database, have fun"));
}
