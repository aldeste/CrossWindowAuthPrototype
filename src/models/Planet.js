// @flow
import { type DataLoaders } from "../schema/apiHelper";
import { type Viewer } from "./User";
import InfoFields, {
  type InfoFieldFromDatabase,
  type InfoFieldTypes
} from "./InfoFields";
import { Planet as dbPlanet } from "../data";

export type PlanetType = InfoFieldTypes & {
  climates: string,
  diameter: number,
  gravity: string,
  name: string,
  orbitalPeriod: number,
  population: number,
  residents: Array<string>,
  rotationPeriod: number,
  surfaceWater: number,
  terrains: string
};

type PlanetModel = InfoFieldFromDatabase & {
  climates: string,
  diameter: number,
  gravity: string,
  name: string,
  orbitalPeriod: number,
  population: number,
  residents: Array<{ id: string }>,
  rotationPeriod: number,
  surfaceWater: number,
  terrains: string
};

export default class PlanetInstance extends InfoFields {
  climates: string;
  diameter: number;
  gravity: string;
  name: string;
  orbitalPeriod: number;
  population: number;
  residents: Array<string>;
  rotationPeriod: number;
  surfaceWater: number;
  terrains: string;

  constructor(data: PlanetModel) {
    super(data);
    this.climates = data.climates;
    this.diameter = data.diameter;
    this.gravity = data.gravity;
    this.name = data.name;
    this.orbitalPeriod = data.orbitalPeriod;
    this.population = data.population;
    this.rotationPeriod = data.rotationPeriod;
    this.surfaceWater = data.surfaceWater;
    this.terrains = data.terrains;
    this.residents = data.residents
      ? data.residents.reduce(
          (previous, current) => [...previous, current.id],
          []
        )
      : [];
  }

  static async gen(
    viewer: ?Viewer,
    id: string,
    { Planet }: DataLoaders
  ): Promise<?PlanetInstance> {
    const data: PlanetModel | null = await Planet.load(id);

    data &&
      console.log(
        data.residents.reduce(
          (previous, current) => [...previous, current.id],
          []
        )
      );

    return data ? new PlanetInstance(data) : null;
  }

  static async genMany(
    viewer: ?Viewer,
    ids: Array<string>,
    { Planet }: DataLoaders
  ): Promise<?Array<PlanetInstance>> {
    const data: Array<PlanetModel> | null = await dbPlanet
      .scope("withIds")
      .findAll({
        where: { id: [ids] }
      });

    // return imideately if failed to fetch
    if (!data) return null;

    // We put each field in the dataloader cache
    await data.forEach(result => Planet.prime(result.id, result));

    return await data.map(result => new PlanetInstance(result));
  }
}
