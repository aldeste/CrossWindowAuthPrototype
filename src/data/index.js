// @flow
import connection from "./databaseConnection";
import { Person, Planet } from "./models";
import generateMockData from "./generateMockData";

/**
 * We define relations here, so they'll be defined before
 * databse is syncronised and included in the export.
 */
Planet.belongsToMany(Person, {
  // constraints: false,
  as: "residents",
  through: "residents"
});

Person.belongsTo(Planet, {
  // constraints: false,
  as: "homeworld"
});

/**
 * Scopes define additional sequelize settings which are
 * callable when fetching from the database. defaultScope
 * is always applied unless another scope is choosen.
 */
Person.addScope(
  "withIds",
  { include: [{ model: Planet, as: "homeworld", attributes: ["id"] }] },
  { override: true }
);

Planet.addScope(
  "withIds",
  { include: [{ model: Person, as: "residents", attributes: ["id"] }] },
  { override: true }
);

export async function initializeDatabase(): Promise<void> {
  await connection.sync({ force: true });
  await generateMockData(true, process.env.NODE_ENV !== "test");
}

export default connection;
export { Person, Planet };
