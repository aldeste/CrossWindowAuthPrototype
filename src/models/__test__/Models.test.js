import Planet from "../Planet";
import User from "../User";
import { Planet as dbPlanet, Person as dbPerson } from "../../data";
import DataLoader from "DataLoader";

function getResolve(type, ids) {
  return Promise.all(ids.map(id => type.scope("withIds").findById(id)));
}

const loaders = {
  Planet: new DataLoader(ids => getResolve(dbPlanet, ids)),
  Person: new DataLoader(ids => getResolve(dbPerson, ids))
};

[User, Planet].map(model =>
  describe(`${model.name} service layer with connection`, () => {
    beforeEach(async () => {
      global.console.log = jest.fn();
      await require("../../data").initializeDatabase();
    });

    it("Returns a class with a valid integer id", async () => {
      const result = await model.gen(null, 3, loaders);
      // Remove created and edited as they're inconsistent
      // after testing if they exist, as they should

      expect(result.created).toBeDefined();
      expect(result.edited).toBeDefined();
      delete result.created;
      delete result.edited;
      expect(result).toMatchSnapshot();
    });

    it("Returns a class with a valid string id", async () => {
      const result = await model.gen(null, "3", loaders);
      // Remove created and edited as they're inconsistent
      // after testing if they exist, as they should

      expect(result.created).toBeDefined();
      expect(result.edited).toBeDefined();
      delete result.created;
      delete result.edited;
      expect(result).toMatchSnapshot();
    });

    it("Returns a class with a valid array of ids", async () => {
      const results = await model.genMany(null, ["1", "2", "3"], loaders);
      // Remove created and edited as they're inconsistent
      // after testing if they exist, as they should

      results.forEach(result => {
        expect(result.created).toBeDefined();
        expect(result.edited).toBeDefined();
        delete result.created;
        delete result.edited;
      });

      expect(results).toMatchSnapshot();
    });

    it("Returns null if id is invalid", async () => {
      const result = await model.gen(null, "This is an invalid id", loaders);
      expect(result).toBeNull();
    });
  })
);
