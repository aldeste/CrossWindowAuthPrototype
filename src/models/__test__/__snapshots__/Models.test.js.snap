// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PlanetInstance service layer with connection Returns a class with a valid array of ids 1`] = `
Array [
  PlanetInstance {
    "GraphQLType": "Planet",
    "climates": "temperate",
    "diameter": 12500,
    "gravity": "1 standard",
    "id": 1,
    "name": "Alderaan",
    "orbitalPeriod": 364,
    "population": 2000000000,
    "residents": Array [
      5,
      65,
      79,
    ],
    "rotationPeriod": 24,
    "surfaceWater": 40,
    "terrains": "grasslands, mountains",
  },
  PlanetInstance {
    "GraphQLType": "Planet",
    "climates": "temperate, tropical",
    "diameter": 10200,
    "gravity": "1 standard",
    "id": 2,
    "name": "Yavin IV",
    "orbitalPeriod": 4818,
    "population": 1000,
    "residents": Array [],
    "rotationPeriod": 24,
    "surfaceWater": 8,
    "terrains": "jungle, rainforests",
  },
  PlanetInstance {
    "GraphQLType": "Planet",
    "climates": "frozen",
    "diameter": 7200,
    "gravity": "1.1 standard",
    "id": 3,
    "name": "Hoth",
    "orbitalPeriod": 549,
    "population": null,
    "residents": Array [],
    "rotationPeriod": 23,
    "surfaceWater": 100,
    "terrains": "tundra, ice caves, mountain ranges",
  },
]
`;

exports[`PlanetInstance service layer with connection Returns a class with a valid integer id 1`] = `
PlanetInstance {
  "GraphQLType": "Planet",
  "climates": "frozen",
  "diameter": 7200,
  "gravity": "1.1 standard",
  "id": 3,
  "name": "Hoth",
  "orbitalPeriod": 549,
  "population": null,
  "residents": Array [],
  "rotationPeriod": 23,
  "surfaceWater": 100,
  "terrains": "tundra, ice caves, mountain ranges",
}
`;

exports[`PlanetInstance service layer with connection Returns a class with a valid string id 1`] = `
PlanetInstance {
  "GraphQLType": "Planet",
  "climates": "frozen",
  "diameter": 7200,
  "gravity": "1.1 standard",
  "id": 3,
  "name": "Hoth",
  "orbitalPeriod": 549,
  "population": null,
  "residents": Array [],
  "rotationPeriod": 23,
  "surfaceWater": 100,
  "terrains": "tundra, ice caves, mountain ranges",
}
`;

exports[`UserInstance service layer with connection Returns a class with a valid array of ids 1`] = `
Array [
  UserInstance {
    "GraphQLType": "Person",
    "birthYear": "19BBY",
    "eyeColor": "blue",
    "gender": "male",
    "hairColor": "blond",
    "height": 172,
    "homeworld": 60,
    "id": 1,
    "mass": 77,
    "name": "Luke Skywalker",
    "skinColor": "fair",
    "token": "cGVvcGxlOjE=",
  },
  UserInstance {
    "GraphQLType": "Person",
    "birthYear": "112BBY",
    "eyeColor": "yellow",
    "gender": "n/a",
    "hairColor": "n/a",
    "height": 167,
    "homeworld": 60,
    "id": 2,
    "mass": 75,
    "name": "C-3PO",
    "skinColor": "gold",
    "token": "cGVvcGxlOjI=",
  },
  UserInstance {
    "GraphQLType": "Person",
    "birthYear": "33BBY",
    "eyeColor": "red",
    "gender": "n/a",
    "hairColor": "n/a",
    "height": 96,
    "homeworld": 7,
    "id": 3,
    "mass": 32,
    "name": "R2-D2",
    "skinColor": "white, blue",
    "token": "cGVvcGxlOjM=",
  },
]
`;

exports[`UserInstance service layer with connection Returns a class with a valid integer id 1`] = `
UserInstance {
  "GraphQLType": "Person",
  "birthYear": "33BBY",
  "eyeColor": "red",
  "gender": "n/a",
  "hairColor": "n/a",
  "height": 96,
  "homeworld": 7,
  "id": 3,
  "mass": 32,
  "name": "R2-D2",
  "skinColor": "white, blue",
  "token": "cGVvcGxlOjM=",
}
`;

exports[`UserInstance service layer with connection Returns a class with a valid string id 1`] = `
UserInstance {
  "GraphQLType": "Person",
  "birthYear": "33BBY",
  "eyeColor": "red",
  "gender": "n/a",
  "hairColor": "n/a",
  "height": 96,
  "homeworld": 7,
  "id": 3,
  "mass": 32,
  "name": "R2-D2",
  "skinColor": "white, blue",
  "token": "cGVvcGxlOjM=",
}
`;
